pkg/antctl/raw/upgrade/apistorage/command.go

```diff
@@ -0,0 +1,263 @@
+		if strings.HasSuffix(crd.Name, "crd.antrea.io") {
+			crdNamesList.Insert(crd.Name)
+		}
+	}
+
+	return crdNamesList, nil
+}
+
+// getCRDNameListToUpgrade is used to get the list of CRDs names to upgrade.
+func getCRDNameListToUpgrade(cmd *cobra.Command, k8sClient client.Client, crdNameListToUpgrade sets.Set[string]) (sets.Set[string], error) {
+	allCRDNameList, err := getAllCRDNames(k8sClient)
+	if err != nil {
+		return nil, err
+	}
+	// If the list of CRDs to upgrade is empty, upgrade all CRDs managed by Antrea.
+	if crdNameListToUpgrade.Len() == 0 {
+		crdNameListToUpgrade = allCRDNameList
+	} else {
+		// If the list of CRDs to upgrade is not empty, and it contains CRDs which are managed by Antrea, skip it.
+		for _, name := range crdNameListToUpgrade.Difference(allCRDNameList) {
+			fmt.Fprintf(cmd.OutOrStdout(), "Skip CRD %q which is not created by Antrea.\n", name)
+		}
+		// Only upgrade the CRDs managed by Antrea.
+		crdNameListToUpgrade = crdNameListToUpgrade.Intersection(allCRDNameList)
+	}
+	return crdNameListToUpgrade, nil
+}
+
+// getCRDListToUpgrade is used to get the list of CRDs to upgrade.
+func getCRDListToUpgrade(cmd *cobra.Command, k8sClient client.Client, crdNameListToUpgrade sets.Set[string]) []*apiextv1.CustomResourceDefinition {
+	var crdsToUpgrade []*apiextv1.CustomResourceDefinition
+	for crdName := range crdNameListToUpgrade {
+		crd := &apiextv1.CustomResourceDefinition{}
+		if err := k8sClient.Get(context.TODO(), client.ObjectKey{Name: crdName}, crd); err != nil {
+			fmt.Fprintf(cmd.ErrOrStderr(), "Failed to get CRD %q: %v.\n", crdName, err)
+			continue
+		}
+		// The CRD that has only one version does not need to be upgraded.
+		if len(crd.Spec.Versions) < 2 {
+			fmt.Fprintf(cmd.OutOrStdout(), "No need to upgrade CRD %q since it only has one version.\n", crdName)
+			continue
+		}
+		crdsToUpgrade = append(crdsToUpgrade, crd)
+	}
+	return crdsToUpgrade
+}
+
+// upgradeCRDObjects is used to upgrade the objects of a CRD.
+func upgradeCRDObjects(cmd *cobra.Command, k8sClient client.Client, crd *apiextv1.CustomResourceDefinition) error {
+	list := &unstructured.UnstructuredList{}
+	list.SetGroupVersionKind(schema.GroupVersionKind{
+		Group:   crd.Spec.Group,
+		Version: getCRDStorageVersion(crd),
+		Kind:    crd.Spec.Names.ListKind,
+	})
+	if err := k8sClient.List(context.TODO(), list); err != nil {
+		return err
+	}
+	if len(list.Items) == 0 {
+		return nil
+	}
+	fmt.Fprintf(cmd.OutOrStdout(), "Upgrading %d CRD %q objects.\n", len(list.Items), crd.Name)
+	for _, obj := range list.Items {
+		copyObj := obj
+		err := retry.OnError(wait.Backoff{
+			Duration: time.Second,
+			Steps:    5,
+		}, func(err error) bool {
+			return handleUpdateErr(err) != nil
+		}, func() error {
+			return k8sClient.Update(context.TODO(), &copyObj)
+		})
+		if handleUpdateErr(err) != nil {
+			return err
+		}
+	}
+	fmt.Fprintf(cmd.OutOrStdout(), "Successfully upgraded %d CRD %s objects.\n", len(list.Items), crd.Name)
```

Comment:
Can you also provide extra info like which version these objects have been upgraded?

Reply:
For example: take CRD named `Demo` as an example:

- Create the CRD with version v1alpha1, the values of `served` and `storage` of the version are both `true`
- Create a version `v1alpha`` `Demo` object `obj-a1` with kubectl, and when we get the object with kubectl, the value of field of `apiVersion` is `v1alpha`
- Update the CRD by adding version v1beta1, the values of `served` and `storage` of the version are both `true`. Note that, update the value of `storage` of version `v1alpha1` to `false`
- Get the object `obj-a1` with kubectl, the value of field of `apiVersion` is still `v1alpha` 
- Create a version `v1beta1` `Demo` object `obj-b1` with kubectl, then we get the objects `obj-a1` and `obj-b1`, the value of field of `apiVersion` is still `v1beta1`
- In the etcd, query these two objects with command `etcdctl get --prefix /registry/crd.antrea.io/demos/<obj name> -w json | jq -r '.kvs[].value' | base64 -d | jq -r '.apiVersion', the output of `obj-a1` is `crd.antrea.io/v1alpha1`, and the output of `obj-b1` is `crd.antrea.io/v1beta1`. 

The issue is that it seems that we cannot get the "real" api version with kubectl or k8s client in code. Could you give some ideas? Thanks! @tnqn @luolanzone

Reply:
Isn't the storage version of the CRD what we need?

Reply:
It seems that I cannot get the original version of a CRD object stored in etcd with k8s clientset, but it can be upgraded by reading and writing it back. BTW, do we need to output the information of the original version of a CRD object when upgrading?

Reply:
I was suggesting to add the info about which old version of the CR and which new version of the CR have been upgraded. The storage version should be the new version, but seems it's difficult to get the old version if the upgrade is done.

Reply:
Yes, I think version of an object can't be fetched via API. It's transparent to clients.